test/ScanDependencies/Inputs/ModuleDependencyGraph.swift

```diff
@@ -13,19 +13,22 @@ import Foundation
 
 enum ModuleDependencyId: Hashable {
   case swift(String)
+  case swiftExternal(String)
   case clang(String)
 
   var moduleName: String {
     switch self {
-      case .swift(let name): return name
-      case .clang(let name): return name
+    case .swift(let name): return name
+    case .swiftExternal(let name): return name
```

Comment:
Another question: why does the client side need to know whether a module is a placeholder dependency? can we blend them in other module dependencies in the JSON output?

Reply:
I wrestled with doing this a little bit, here is my rationale for making it a distinct dependency kind: 

Using `swift-driver` as an example, when the scanner produces a dependency graph for the `swift-driver`, it is not enough to have the placeholder just be another swift module in it. One of the things the client will also need (and will be supplied by SwiftPM in this case), is a full dependency graph of the placeholder itself, to get its transitive dependencies and to _unique_ all such transitive dependencies with the modules already in the dependency graph of the main module. Therefore, once the scanner produces this graph, the client will go through all placeholders and substitute their leaf nodes in this graph with a full dependency tree, replacing placeholders with "actual" dependencies.

It made sense to model such nodes as a distinct kind of entity in the graph instead of relying on the client searching for placeholders by-name, but I am curious what you think.

Reply:
> go through all placeholders and substitute their leaf nodes in this graph with a full dependency tree
If we need to transform the graph any way, can we just scan at the point when actual dependencies are ready? avoiding placeholder dependencies altogether?

Reply:
This particular aspect of this design comes from the way SwiftPM (and I suspect other build systems?) interact with the compiler/driver. SwiftPM forms a build manifest for LLBuild ahead of time, for all targets; instead of planning & building a specific target, then requesting a build plan/manifest for the next target, etc. So this approach is meant to fit with that model. Hence, the driver needs to plan a build for a target/module when its dependencies have not yet been built. 

I am also not sure how we would handle transitive dependencies in that case. If we have a target that is a dependency that has been pre-built, I think we would still need its dependency graph to get transitive dependency modules?

Reply:
An important point to keeping SwiftPM's design w.r.t. planning a full package manifest ahead of time for all targets maximizes the parallelism we can extract from the full package build graph. Building targets one-by-one would serialize this, to a degree.

Reply:
It makes a lot of sense now. Thank you, Artem! Could you document this in the source code explaining the post-processing steps clients will do for these placeholder dependencies and why they are necessary?

Reply:
Will do!

Reply:
btw, the post-processing steps make the term placeholder more proper :)

Reply:
> btw, the post-processing steps make the term placeholder more proper :)

It also lends to nicer APIs in the client, `resolvePlaceholders` is a very natural-sounding operation.

Reply:
We actually don't need the synthesized module path for these placeholders, right? we could use some internal ID to facilitate find and replacement. But it's minor though.